Source/WTF/wtf/glib/SysprofAnnotator.h

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright (C) 2024 Igalia, S.L.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Library General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Library General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Library General Public License
+ *  along with this library; see the file COPYING.LIB.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#include <cstdio>
+#include <sysprof-capture.h>
+#include <wtf/HashMap.h>
+#include <wtf/Lock.h>
+#include <wtf/Vector.h>
+#include <wtf/text/ASCIILiteral.h>
+#include <wtf/text/CString.h>
+
+namespace WTF {
+
+class SysprofAnnotator final {
+    WTF_MAKE_NONCOPYABLE(SysprofAnnotator);
+
+    using RawPointerPair = std::pair<const void*, const void*>;
+    using TimestampAndString = std::pair<int64_t, Vector<char>>;
+
+public:
+
+    static SysprofAnnotator& singleton()
+    {
+        static SysprofAnnotator instance;
+        return instance;
+    }
+
+    void instantMark(const char *name, const char *description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, 0, m_processName, name, description, args);
+        va_end(args);
+    }
+
+    void mark(Seconds timeDelta, const char *name, const char *description, ...) WTF_ATTRIBUTE_PRINTF(4, 5)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, timeDelta.microsecondsAs<int64_t>(), m_processName, name, description, args);
+        va_end(args);
+    }
+
+    void beginMark(const void* pointer, const char* name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(4, 5)
+    {
+        auto key = std::make_pair(pointer, static_cast<const void*>(name));
+
+        Vector<char, 1024> buffer;
```

Comment:
Use std::array here. Vector is for when you need to change the size of the buffer, but here you're using a fixed size, so it's not needed.

I suppose this goes for TimestampAndString as well. Normally I would ask you to change TimestampAndString to use a normal String instead of this handrolled string, but I see Carlos requested it be done this way to avoid an extra copy. So I suppose he's going to prefer std::array. Whatever.

Reply:
Apparently std::array conflicts with HashMap, it complains about the size being too big and it's too inefficient to move when rehashing.

Reply:
Oh good point! That's reason enough to prefer Vector.